### PR TITLE
add mix hex.owner packages task

### PR DESCRIPTION
**Pull request by fteem**
Related to https://github.com/hexpm/hex/issues/147 
Also, this requires the changes made in https://github.com/hexpm/hex_web/pull/183

### CODE REVIEW

File: lib/mix/tasks/hex/owner.ex [hunk 1]
```diff
@@ -81,4 +90,33 @@ defmodule Mix.Tasks.Hex.Owner do
...
   end
+
+  def list_owned_packages do
+    config = Hex.Config.read
+    username = local_user(config)
+    auth = Utils.auth_info(config)
+
+    case Hex.API.User.get(username, auth) do
+      {code, body} when code in 200..299 ->
+        Hex.Shell.info "Owned packages:"
```
Comment (ericmj) [inline lib/mix/tasks/hex/owner.ex]: We don't need to print this, it is already implied.

File: lib/mix/tasks/hex/owner.ex [hunk 2]
```diff
@@ -81,4 +90,33 @@ defmodule Mix.Tasks.Hex.Owner do
...
+        body["owned_packages"]
+        |> Map.to_list
```
Comment (ericmj) [inline lib/mix/tasks/hex/owner.ex]: This line is redundant.

File: lib/mix/tasks/hex/owner.ex [hunk 3]
```diff
@@ -81,4 +90,33 @@ defmodule Mix.Tasks.Hex.Owner do
...
+    end
+  end
+
+  defp local_user(config) do
+    case Keyword.fetch(config, :username) do
+      {:ok, username} ->
+        username
+      :error ->
+        Mix.raise "No user authorised on the local machine. Run `mix hex.user auth` " <>
+                  "or create a new user with `mix hex.user register`"
```
Comment (ericmj) [inline lib/mix/tasks/hex/owner.ex]: We don't need to handle this case, `Utils.auth_info` should already have raised.

File: lib/mix/tasks/hex/owner.ex [hunk 4]
```diff
@@ -81,4 +90,33 @@ defmodule Mix.Tasks.Hex.Owner do
...
+    case Hex.API.User.get(username, auth) do
+      {code, body} when code in 200..299 ->
+        Hex.Shell.info "Owned packages:"
+        body["owned_packages"]
+        |> Map.to_list
+        |> Enum.each(fn({name, url}) ->
+          Hex.Shell.info("#{name} - #{url}")
+        end)
+      {code, body} ->
+        Hex.Shell.error("Failed to auth")
```
Comment (ericmj) [inline lib/mix/tasks/hex/owner.ex]: We don't know the failure reason, please change to "Listing owned packages failed" instead.

File: lib/mix/tasks/hex/owner.ex [hunk 5]
```diff
@@ -81,4 +90,33 @@ defmodule Mix.Tasks.Hex.Owner do
         Hex.Utils.print_error_result(code, body)
     end
   end
+
+  def list_owned_packages do
+    config = Hex.Config.read
```
Comment (ericmj) [inline lib/mix/tasks/hex/owner.ex]: We should avoid reading the config multiple times, instead pass the config from the caller where we also read it.

File: lib/mix/tasks/hex/owner.ex [hunk 6]
```diff
@@ -81,4 +90,19 @@ defmodule Mix.Tasks.Hex.Owner do
         Hex.Utils.print_error_result(code, body)
     end
   end
+
+  def list_owned_packages(auth) do
+    config = Hex.Config.read
```
Comment (ericmj) [inline lib/mix/tasks/hex/owner.ex]: Can you pass the config from the caller, we should only read it once.

**Comment by ericmj** [comment]
I'm not sure if it should be called as `mix hex.owner packages` or something else like `mix hex.owner list` or maybe `mix hex.owner owned_packages`.

WDYT? /cc @lexmag @josevalim

**Comment by fteem** [comment]
I updated the PR - is this what you were aiming at?
Regarding the naming - I am not sure. Wouldn't go with `mix hex.owner list` because it might be confused with the `mix hex.owner list PACKAGE` task. Also, the naming in `mix hex.owner owned_packages` seems a bit redundant IMHO.

**Comment by ericmj** [comment]
Okay, left one comment, squash then we are good to merge.

**Comment by fteem** [comment]
@ericmj done, thanks for the help :)

**Comment by lexmag** [comment]
@ericmj I think `mix hex.owner packages` is the best fit here, can't come up with anything better.

**Comment by ericmj** [comment]
Perfect, thanks!

**Pull request merged**